File: src_fortran/sub/layerhit.f
c
c
c
c
c
c
      subroutine LAYERHIT(rp,nrpdeep,nl,nrp,mll)
c
c     counts the number of hits in each layer.
c
      implicit none
      include '../inc/vel_com.inc'
c
      integer nrpdeep,nl,nrp,mll
      real rp(3,inrpmax)
      real rpmax,avraydepth
      integer j,jlay
c
c
c     find deepest layer hit by this ray (= refraction layer or eventlayer
c                                           for direct wave)
      rpmax=-999.
      do j=1,nrp
         if(rp(3,j).gt.rpmax)then
            rpmax=rp(3,j)  ! = max. depth of ray
            nrpdeep=j   ! index of deepest raypoint
         endif
      enddo
      rpmax=rpmax+0.000001   ! avoid raypoint on layer-boundary
      lmax=1
      do j=2,nl
         if(h(j).gt.rpmax)then   !  h(j)= top of layer j [km]
            lmax=j-1  ! lmax = layer number for depth rpmax
            goto 4000
         endif
      enddo
      lmax=nl  ! lowest layer
 4000 continue
      if(rp(3,nrpdeep).eq.rp(3,nrpdeep+1))then   ! ray is horizontal
         irefrlayer(lmax)=irefrlayer(lmax)+1       ! headwave in layer LMAX
         refraylen(lmax)=refraylen(lmax)+
     &   SQRT( (rp(1,nrpdeep)-rp(1,nrpdeep+1))**2 +
     &         (rp(2,nrpdeep)-rp(2,nrpdeep+1))**2 )
      else
         if(mll.eq.0)then
            noheadwave=noheadwave+1
         else
            irefllayer(lmax-1)=irefllayer(lmax-1)+1
c                                           ! -1 because of rpmax=rpmax+0.000001
         endif
      endif
      avhraylen=avhraylen+
     & SQRT( (rp(1,1)-rp(1,nrp))**2 +
     &       (rp(2,1)-rp(2,nrp))**2 )
      avvraylen=avvraylen+
     & ABS( rp(3,nrpdeep)-rp(3,nrp) )
c
      do j=2,nrp
         avraydepth=( rp(3,j)+rp(3,j-1) ) / 2.
         avraydepth=avraydepth+0.000001    ! avoid raydepth on layer-boundary
         do jlay=1,nl
            if(avraydepth.ge.h(jlay).and.
     &         avraydepth.lt.(h(jlay)+thk(jlay)) )then
               hitlay(jlay,1)=hitlay(jlay,1)+1
               hitlay(jlay,2)=hitlay(jlay,2)+
     &                        SQRT( (rp(1,j)-rp(1,j-1))**2 +
     &                              (rp(2,j)-rp(2,j-1))**2 )
               hitlay(jlay,3)=hitlay(jlay,3)+
     &                        ABS( rp(3,j)-rp(3,j-1) )
            endif
         enddo
      enddo
      RETURN
      end ! of subr. layerhit

File: src/sub/layerhit.c
/*
 * Count ray hits in each layer and compute ray path statistics
 * from Fortran LAYERHIT subroutine
 */

#include <stddef.h>
#include <math.h>
#include "../include/globals.h"

/* External global variables from vel_com.inc */
extern int lmax;
extern int noheadwave;
extern double avhraylen;
extern double avvraylen;
extern int irefllayer[];
extern int irefrlayer[];
extern double refraylen[];
extern double hitlay[][3];
extern double h[];      /* Layer top depths */
extern double thk[];    /* Layer thicknesses */

void layerhit(double rp[3][200], int *nrpdeep, int nl, int nrp, int mll) {
    double rpmax, avraydepth;
    int j, jlay;
    
    /* Find deepest ray point (maximum depth) */
    rpmax = -999.0f;
    for (j = 0; j < nrp; j++) {
        if (rp[2][j] > rpmax) {
            rpmax = rp[2][j];
            *nrpdeep = j;  /* Index of deepest ray point (0-based) */
        }
    }
    
    rpmax = rpmax + 0.000001f;  /* Avoid exact layer boundary */
    
    /* Find deepest layer hit by this ray */
    lmax = 0;  /* Layer 0 by default */
    for (j = 1; j < nl; j++) {
        if (h[j] > rpmax) {
            lmax = j - 1;  /* Layer number (0-based) for depth rpmax */
            break;
        }
    }
    if (lmax == 0 && h[1] <= rpmax) {
        lmax = nl - 1;  /* Lowest layer if depth exceeds all layer tops */
    }
    
    /* Check if ray is horizontal (headwave) */
    if (*nrpdeep >= 0 && *nrpdeep < nrp - 1 &&
        rp[2][*nrpdeep] == rp[2][*nrpdeep + 1]) {
        irefrlayer[lmax]++;  /* Count refraction in layer LMAX */
        double hz_dist = sqrtf((rp[0][*nrpdeep] - rp[0][*nrpdeep + 1]) *
                              (rp[0][*nrpdeep] - rp[0][*nrpdeep + 1]) +
                              (rp[1][*nrpdeep] - rp[1][*nrpdeep + 1]) *
                              (rp[1][*nrpdeep] - rp[1][*nrpdeep + 1]));
        refraylen[lmax] = refraylen[lmax] + hz_dist;
    } else {
        /* Non-horizontal ray */
        if (mll == 0) {
            noheadwave++;
        } else if (lmax > 0) {
            irefllayer[lmax - 1]++;  /* Count reflection in layer above */
        }
    }
    
    /* Accumulate horizontal and vertical ray path lengths */
    double hz_tot = sqrtf((rp[0][0] - rp[0][nrp - 1]) * (rp[0][0] - rp[0][nrp - 1]) +
                         (rp[1][0] - rp[1][nrp - 1]) * (rp[1][0] - rp[1][nrp - 1]));
    avhraylen = avhraylen + hz_tot;
    
    double vz_tot = fabs(rp[2][*nrpdeep] - rp[2][nrp - 1]);
    avvraylen = avvraylen + vz_tot;
    
    /* Accumulate hit statistics for each ray segment */
    for (j = 1; j < nrp; j++) {
        avraydepth = (rp[2][j] + rp[2][j - 1]) / 2.0f;
        avraydepth = avraydepth + 0.000001f;  /* Avoid exact layer boundary */
        
        for (jlay = 0; jlay < nl; jlay++) {
            /* Check if ray segment is in layer jlay */
            if (avraydepth >= h[jlay] && avraydepth < (h[jlay] + thk[jlay])) {
                hitlay[jlay][0]++;  /* Count hits */
                
                double seg_hz = sqrtf((rp[0][j] - rp[0][j - 1]) *
                                     (rp[0][j] - rp[0][j - 1]) +
                                     (rp[1][j] - rp[1][j - 1]) *
                                     (rp[1][j] - rp[1][j - 1]));
                hitlay[jlay][1] = hitlay[jlay][1] + seg_hz;  /* Horizontal distance */
                
                double seg_vz = fabs(rp[2][j] - rp[2][j - 1]);
                hitlay[jlay][2] = hitlay[jlay][2] + seg_vz;  /* Vertical distance */
            }
        }
    }
}
